### ISSUE TITLE
GroupKFold crossvalidation in AutoGluon Tabular?

### ISSUE THREAD

**Issue by davidabek1** [awslabs/autogluon#901]
Hello, 
Thanks for your work on this AutoML package, 
**Is GroupKFold-cross-validation supported in AutoGluon Tabular?**
As AutoGluon creating validation sets for internal search of best model and hyperparameters, it has to create the train/val sets based on allowed groups, or training will be "contaminated".
currently with other models I'm using nested cross-validation approach, where in my inner cv using randomized/baysian parameters search, the lines of code considering the groups indices:
```
cv_inner = GroupKFold(n_splits=3)
clf = RandomizedSearchCV(model, param_distributions=param_dist, cv=cv_inner, n_iter=200, scoring='accuracy')
search = clf.fit(X_train, y_train, groups=groups_train)
```
Thank you.

**Comment by Innixma** [comment]
This is not supported at present, but has been requested before and is a reasonable enhancement to add.

AutoGluon currently supports KFold splits by the following code:

```
def generate_kfold(X, y=None, n_splits=5, random_state=0, stratified=False, n_repeats=1):
    if stratified and (y is not None):
        if n_repeats > 1:
            kf = RepeatedStratifiedKFold(n_splits=n_splits, n_repeats=n_repeats, random_state=random_state)
        else:
            kf = StratifiedKFold(n_splits=n_splits, shuffle=True, random_state=random_state)

        kf.get_n_splits(X, y)
        return [[train_index, test_index] for train_index, test_index in kf.split(X, y)]
    else:
        if n_repeats > 1:
            kf = RepeatedKFold(n_splits=n_splits, n_repeats=n_repeats, random_state=random_state)
        else:
            kf = KFold(n_splits=n_splits, shuffle=True, random_state=random_state)

        kf.get_n_splits(X)
        return [[train_index, test_index] for train_index, test_index in kf.split(X)]
```

Note that `n_repeats` allows for repeated bagging / cv. It seems that `GroupKFold` would not support `n_repeats>1`, so we would need to force `n_repeats=1` in this scenario. Similarly, `stratified` and `random_state` arguments would be ignored.

Beyond updates to this function, we will need to add an additional argument to `TabularPredictor.fit` which allows the passing of the `groups` argument downstream, and correctly update it when we perform actions such as data augmentation / dropping rows.

I will plan to look into this when I find time, as it shouldn't be overly difficult to add.

**Comment by davidabek1** [comment]
Thank you, for the fast response, 
I did looked up to see previous request, and closest resembled one was the "Time based crossvalidation", which is not the same.
All your comments for GroupKFold constraints are in place, another approach as an addition enhancement to provide support for `random_state` and `n_repeats` is with the [GroupShuffleSplit](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GroupShuffleSplit.html#sklearn.model_selection.GroupShuffleSplit).
Thank you, hope this enhancement will be soon implemented.

**Comment by davidabek1** [comment]
@Innixma , thank you for adding this enhancment to the task list of 0.3 release.
Lately I've been playing with H2O and by just adding the ability to assign a fold column, can resolve all posibilities, 
Since then I will be able to create the column to which ever fold split I need, taking care of stratification and/or groups and/or TimeSeries.
All what left for the model is to use the cross-validation only according to assigned folds in this column, the only Cons I can think of is that it requires higher understanding and some nice examples can handle this, I'll put my used example:
```
cv_outer = GroupKFold(n_splits=4)
for train_index, test_index in cv_outer.split(X, y, groups):
    X_train, X_test = X.iloc[train_index, :], X.iloc[test_index, :]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]
    groups_train, groups_test = groups.iloc[train_index], groups.iloc[test_index]

    cv_inner = GroupKFold(n_splits=3)
    for fold, (in_index_ oof_index) in enumerate(cv_inner.split(X_train, y_train, groups)):
        X_train.loc[X_train.index[oof_index], 'fold'] = fold

```
The inner cross-validation will assign correctly (considering groups) fold number to each of the rows, using it for cross validation,
inside the model will output expected results.
The same can be done by manipulating the 'fold' column to any of existing supported spliting schemes in SKlearn.

The outer cross-validation is just to create an actual holdouts test sets, again I have to take care of groups and prevent data leakage.

**Comment by Innixma** [comment]
LeaveOneGroupOut support has been added in #1224.

To use it, refer to the documentation of the `groups` parameter to `TabularPredictor`, available in v0.3 release.

Thanks @davidabek1 for your very helpful comments and suggestions in this thread!
